source/extensions/tracers/zipkin/zipkin_core_types.cc

```diff
@@ -219,6 +238,41 @@ const std::string Span::toJson() {
   return json_string;
 }
 
+const zipkin::proto3::Span Span::toProtoSpan() const {
+  zipkin::proto3::Span span;
+  span.set_trace_id(traceIdAsByteString());
+  span.set_parent_id(parentIdAsByteString());
+  span.set_id(idAsByteString());
+  span.set_name(name_);
+
+  if (timestamp_) {
+    span.set_timestamp(timestamp_.value());
+  }
+
+  if (duration_) {
+    span.set_duration(duration_.value());
+  }
+
+  for (const auto& annotation : annotations_) {
+    if (annotation.isSetEndpoint()) {
+      if (annotation.value() == ZipkinCoreConstants::get().CLIENT_SEND) {
+        span.set_kind(zipkin::proto3::Span::CLIENT);
+        span.mutable_local_endpoint()->MergeFrom(annotation.endpoint().toProtoEndpoint());
+      } else if (annotation.value() == ZipkinCoreConstants::get().SERVER_RECV) {
+        span.set_kind(zipkin::proto3::Span::SERVER);
+        span.mutable_remote_endpoint()->MergeFrom(annotation.endpoint().toProtoEndpoint());
```

Comment:
this is a bug. both the endpoint of client and server annotations is the local endpoint as what is local is discussing this process. There's a special v1 annotation "sa" and "ca" which is about the remote endpoint.

In general it is easier to take code that is v2 and make the data v1 or v2 vs take code that is v1 and make data v1 or v2 because v1 concepts are awkward.

Reply:
Gotcha. Sorry for not realizing it in the first place. Will _re-build_ the models then.